### ISSUE TITLE
Allow for a way to ensure some source generators run before / after others

### ISSUE THREAD

**Issue by pranavkm** [dotnet/roslyn#57239]
We briefly discussed this during 6.0 / C# 10 but ultimately punted it since the ask came in too late. We've had issues where user-written source generators can no longer observe the output of source generated by Razor's:

* https://github.com/dotnet/aspnetcore/issues/32253
* https://github.com/dotnet/aspnetcore/issues/32949
* https://github.com/dotnet/aspnetcore/issues/32172

For 6, the solution we gave our users is to turn off the source generator and fall back to our 2-phase compilation process. We'd made some progress discussing this internally, but filing this for C# 11 consideration.

**Comment by bugproof** [comment]
I really want to use this feature with Blazor but seems like I have to rely on AOP frameworks instead.

**Comment by jcouv** [comment]
Another instance of this problem which was [reported](https://github.com/dotnet/roslyn/issues/59780) is using RegexGenerator in Blazor project.

**Comment by MarkFl12** [comment]
Yeah, I've been trying to build a source generator to help with Blazor projects, but I think what's happening is that it needs to run before the generator that transforms the .razor files so that when their generator runs the code I've generated that's used in them is available

**Comment by RyoukoKonpaku** [comment]
+1 for this. 

We've been using generators for Blazor to generate a strongly typed class that contains all routes of the application so routing isn't stringly typed and much less prone to human errors. This relies on us having to us finding route attributes on the generated files by the razor compiler. Current workaround for us is to define the `RouteAttribute` in a codebehind so our source generator can see it, but ideally it'd be nice to keep using the `@page` directive for this.

**Comment by Blackclaws** [comment]
I think we might want to get some traction or at least ideas going on how this might be be achieved.

In general I see that there are _at least_ two ways this could work. 

1. Re-run generators on the output of all other generators until a steady state is achieved

This would probably be the easiest way to implement this. However this already sounds like a _really_ bad idea in case you build two generators that always trigger each other, leading to an infinite loop. This would however allow generators to use each other, with the possibility of terminating after a maximum amount of round trips.

2. Actually order generators properly.

The way this could work is by having introducing for example an interface
```
public interface IOrderedGenerator
{
   IEnumerable<string> GetBeforeGenerators();
   IEnumerable<string> GetAfterGenerators();
}
```
With the returned strings understood as assembly names.

This would allow a generator to declare both before and after targets. 
- Before generators is relevant for generators that themselves generate attributes that are understood by other generators. 
- After generators make sense in the context of generators like the Razor Source generator that create code from additional files

This method would also make it easy to detect loops, allow parallel execution of generators that aren't part of the same dependency graph, etc.

**Comment by sanchez** [comment]
Sorry bit new to the thread but wouldn't it be possible to just take the order of execution as a list, set in the running csproj? Could just be a bunch of namespaces etc. Then the responsibility is on the consumer and developer documentation.

Additionally different libraries could read in this list and "validate" their ordering and return warnings if it's not as expected.

What I don't like about my suggestion is that it's polluting the csproj more, but to me it makes sense; if I wanted to know/manage my source generators one of the first places I would look is the csproj

**Comment by Youssef1313** [comment]
> Sorry bit new to the thread but wouldn't it be possible to just take the order of execution as a list, set in the running csproj? Could just be a bunch of namespaces etc. Then the responsibility is on the consumer and developer documentation.
> 
> Additionally different libraries could read in this list and "validate" their ordering and return warnings if it's not as expected.
> 
> What I don't like about my suggestion is that it's polluting the csproj more, but to me it makes sense; if I wanted to know/manage my source generators one of the first places I would look is the csproj

This will prevent possible parallelization, I think

Suppose `GeneratorA` should run after `Generator B`, but `GeneratorC` doesn't have any dependency.
It should be possible to start running both `GeneratorA` and `GeneratorC` in parallel.

**Comment by sanchez** [comment]
True. I hadn't considered that. You could group/nest sources thought right?

```xml
<Generators>
  <Step>
    <Source Type="GeneratorA" />
    <Source Type="GeneratorB" />
  </Step>
  <Step>
    <Source Type="GeneratorC" />
  </Step>
</Generators>
```

Each step being parallel within but being sequential adjacent. So A and B parallel together and C after

**Comment by Blackclaws** [comment]
> Sorry bit new to the thread but wouldn't it be possible to just take the order of execution as a list, set in the running csproj? Could just be a bunch of namespaces etc. Then the responsibility is on the consumer and developer documentation.
> 
> Additionally different libraries could read in this list and "validate" their ordering and return warnings if it's not as expected.
> 
> What I don't like about my suggestion is that it's polluting the csproj more, but to me it makes sense; if I wanted to know/manage my source generators one of the first places I would look is the csproj

The problem I think is that this firmly puts the burden on the developer _using_ the generator to know what it generates and how the dependencies between generators might play out.

I think the pit of success is rather that generators themselves define _if_ they handle output from other generators _or_ generate output that other generators (possibly dependencies) consume. A developer wouldn't really want to worry about these specific details and just expects to be able to use source generators in a way that they can build on one another without needing an extra step.

Also this would introduce another point where this needs to be introduced.

**Comment by CyrusNajmabadi** [comment]
I'm not seeing any mechanism for this sort of system to be fast enough. Just having a single level of generators today, is enormously impactful for perf. Even for things that used to be fast (like creating a ref-assembly) we now have to fully bind and create the initial compilation so we can run generators. If we then allow generators to run before/after, and we expect those generators to understand the semantics of that prior generated code, I do not see how we could do that without just huge perf hits.

**Comment by juarezasjunior** [comment]
I liked the idea of the dev specify the order in .csproj. But, sometimes it could be boring, special if the dev is using a package that he doesn't know that it has code being generated.
What about if we have both options? The default behavior is the same that we have now. The projects in .csproj will execute without order. But, if we want to execute it following an order, we may specify the order in .csproj like mentioned by @sanchez https://github.com/dotnet/roslyn/issues/57239#issuecomment-1238160592?

**Comment by phil-scott-78** [comment]
I'm not sure ordering would ever be possible. Take the Razor generator. There are a decent amount of generators that would need to look at the code in the razor files while also letting those razor files consume the generated code. I think that leaves multi-pass runs as the only option.

I suspect, maybe with hopeless optimism, that if this multi-pass behavior was restricted to incremental generators that the output could be examined rather quickly to determine if more runs are required, with a failsafe to bail if recursion is detected.

**Comment by Blackclaws** [comment]
> I'm not seeing any mechanism for this sort of system to be fast enough. Just having a single level of generators today, is enormously impactful for perf. Even for things that used to be fast (like creating a ref-assembly) we now have to fully bind and create the initial compilation so we can run generators. If we then allow generators to run before/after, and we expect those generators to understand the semantics of that prior generated code, I do not see how we could do that without just huge perf hits.

Performance is secondary _kind of_. This is a feature that might have a decent performance hit, but as long as its something that generators themselves have to implement in order for that to matter I think the problem isn't that big.

> I liked the idea of the dev specify the order in .csproj. But, sometimes it could be boring, special if the dev is using a package that he doesn't know that it has code being generated. What about if we have both options? The default behavior is the same that we have now. The projects in .csproj will execute without order. But, if we want to execute it following an order, we may specify the order in .csproj like mentioned by @sanchez [#57239 (comment)](https://github.com/dotnet/roslyn/issues/57239#issuecomment-1238160592)?

I think specifically that the default option being no order doesn't make a lot of sense. If I already know as the author of a generator that it will generate output that another generator has to process I want to be able to declare this in advance.

> I'm not sure ordering would ever be possible. Take the Razor generator. There are a decent amount of generators that would need to look at the code in the razor files while also letting those razor files consume the generated code. I think that leaves multi-pass runs as the only option.
> 
> I suspect, maybe with hopeless optimism, that if this multi-pass behavior was restricted to incremental generators that the output could be examined rather quickly to determine if more runs are required, with a failsafe to bail if recursion is detected.

I think for razor files specifically this is a difficult situation. Looking at the generated code, I don't think there is a reason why Razor itself would have this sort of circular dependency on the output of the source generator. It might be there and since this is definitely one of the most visible use cases for chaining generators we definitely have to think about it though.

The razor source generator might itself be split into two parts though. The first part generates a file that just has all the attributes and declares a partial class, and the second part implements the actual rendering pipeline. That way you could hook into the middle of the razor generator.

**Comment by CyrusNajmabadi** [comment]
Performance needs to be primary. The perf here affects literally every semantic task at a higher layer.

**Comment by Blackclaws** [comment]
> Performance needs to be primary. The perf here affects literally every semantic task at a higher layer.

Sure, I'm not saying its completely irrelevant, I'm saying that as long as the behavior is opt in, it shouldn't have any impact on current compile times. When there is a generator that _needs_ this behavior its secondary how performant it is, as long as it isn't abysmal.

**Comment by CyrusNajmabadi** [comment]
> Sure, I'm not saying its completely irrelevant, I'm saying that as long as the behavior is opt in, it shouldn't have any impact on current compile times.

**Comment by CyrusNajmabadi** [comment]
> Ah, the multi pass scenario is what I consider a bad solution though. I'd much rather have a directed acyclic graph of generators so that only generators that actually depend on one another have to be run in sequence. 

This is still problematic.  Just consider the simple case of two generators, where one depends on the other.  To actually produce semantics we ahve to do:

1. Produce initial `Compilation-A`.
3. Run generator 'G1' on it to generate `Sources-A'`
4. Produce `Compilation-B` by combining `Sources-A'` to `Compilation-A`
5. Run generator 'G2' on it to generate `Sources-B'`.
6. Produce `Compilation-C` by combining `Sources-B'` to `Compilation-B`.
7. Compile and emit `Compilation-C`

This is a *ton* of very expensive work three *full* compilations are made here.

**Comment by Blackclaws** [comment]
> > Ah, the multi pass scenario is what I consider a bad solution though. I'd much rather have a directed acyclic graph of generators so that only generators that actually depend on one another have to be run in sequence.
> 
> This is still problematic. Just consider the simple case of two generators, where one depends on the other. To actually produce semantics we ahve to do:
> 
>     1. Produce initial `Compilation-A`.
> 
>     2. Run generator 'G1' on it to generate `Sources-A'`
> 
>     3. Produce `Compilation-B` by combining `Sources-A'` to `Compilation-A`
> 
>     4. Run generator 'G2' on it to generate `Sources-B'`.
> 
>     5. Produce `Compilation-C` by combining `Sources-B'` to `Compilation-B`.
> 
>     6. Compile and emit `Compilation-C`
> 
> 
> This is a _ton_ of very expensive work three _full_ compilations are made here.

Right. I'm not saying I have to answer yet on _how_ this can be achieved with good performance, just that it is a feature that's been requested again and again and that, with increasing proliferation of source generators, will only add more value.

I'm wondering whether the compilation that is passed to a second generator really has to be a _full_ compilation or whether there are ways to make this interface more lightweight. 

Also my hope would be that incremental generators also reduce the overall amount of calls that need to be made to the source generators. This dependency feature is something I firmly believe shouldn't be added to the old V1 source generators.

**Comment by CyrusNajmabadi** [comment]
> Right. I'm not saying I have to answer yet on how this can be achieved with good performance, just that it is a feature that's been requested again and again and that, with increasing proliferation of source generators, will only add more value.

I get teh value side.  But the value is only acceptable if it doesn't come with even more negative value from teh perf hit there.  As before, this is not speculative.  We're *still* working on the perf problems introduced by generators (including 'incremental generators).  We're not even close to back to where we want to be, and that's the *current* state of things.  We can't layer on even more perf sinks when our perf is not even at an acceptable level today.

> Also my hope would be that incremental generators also reduce the overall amount of calls that need to be made to the source generators.

**Comment by Blackclaws** [comment]
> > Right. I'm not saying I have to answer yet on how this can be achieved with good performance, just that it is a feature that's been requested again and again and that, with increasing proliferation of source generators, will only add more value.
> 
> I get teh value side. But the value is only acceptable if it doesn't come with even more negative value from teh perf hit there. As before, this is not speculative. We're _still_ working on the perf problems introduced by generators (including 'incremental generators). We're not even close to back to where we want to be, and that's the _current_ state of things. We can't layer on even more perf sinks when our perf is not even at an acceptable level today.

I think that depends on what you deem an acceptable level. I think we can agree however that we want this feature to be as performant as possible. Introducing this feature would introduce a performance hit on certain critical paths, but I don't really see a way to shorten that path either.

> 
> > Also my hope would be that incremental generators also reduce the overall amount of calls that need to be made to the source generators.
> 
> Not really sure hwo that would work. But if you can come up with some mechanism to make that possible, that would be interesting.

There are a lot of things you can do with incremental generators that would minimize the impact of implementing this feature.

For example you wouldn't rerun dependent generators if their input didn't change, and there is nothing stopping you from including their output from the last compilation if their output won't change.

There's also the point that you _could_ run all generators in parallel in the first iteration, then check if any of the generators would have their pipelines triggered and then selectively rerun those, discarding their previous output and creating a new compilation. My gut feeling is that there are a lot of things you can do, but it won't be easy.

**Comment by phil-scott-78** [comment]
*What if* there was an additional interface that a generator could implement in addition to `IIncrementalGenerator` that would be given the syntax trees of the code that was added by other source generators for that pass. Given that info the author could then opt-in for additional runs. Ideally the new method would have the ability to cache the pipelines just like the incremental generator, and secondary runs would likewise still have the ability to cache their pipelines too.

**Comment by Blackclaws** [comment]
> _What if_ there was an additional interface that a generator could implement in addition to `IIncrementalGenerator` that would be given the syntax trees of the code that was added by other source generators for that pass. Given that info the author could then opt-in for additional runs. Ideally the new method would have the ability to cache the pipelines just like the incremental generator, and secondary runs would likewise still have the ability to cache their pipelines too.

Interesting idea, you'd run into the problem that you can't force one generator to run before another one though. So support for handling other generators would have to be built into new generators.

**Comment by phil-scott-78** [comment]
> Interesting idea, you'd run into the problem that you can't force one generator to run before another one though. So, support for handling other generators would have to be built into new generators.

I'm thinking the generators with the interface would eventually get the output from the other generators in subsequent runs. The method would just keep getting called until generators all agree they don't need to be rerun (or a failsafe tells them all to move it along). 

Generators not implementing the method would be at the whims of Roslyn when they get run, which wouldn't be any worse than the current state.

**Comment by CyrusNajmabadi** [comment]
> The method would just keep getting called until generators all agree they don't need to be rerun 

I don't see any viable path forward on that.  Just running things until a hopeful fixed-point is reached seems like a recipe for disaster (Even with any sort of cap).  If things have circular dependencies, they need to be redesigned to not be that way.

**Comment by phil-scott-78** [comment]
Kinda feels either the design allows two generators to work off each others outputs with a fail safe to keep people from being ridiculous, or a strictly ordered system with a design that the Razor generator would almost immediately throw a wrench in.

**Comment by CyrusNajmabadi** [comment]
>  and creating a new compilation. My gut feeling is that there are a lot of things you can do, but it won't be easy.

Any new compilation creates a linear additional cost on the entire pipeline.  e.g. it's as if we're just recompiling the project 'one more time'.

**Comment by Blackclaws** [comment]
> > The method would just keep getting called until generators all agree they don't need to be rerun
> 
> I don't see any viable path forward on that. Just running things until a hopeful fixed-point is reached seems like a recipe for disaster (Even with any sort of cap). If things have circular dependencies, they need to be redesigned to not be that way.

I agree that circular dependencies are probably not going to work as they are evil (TM) by nature and at least for a first implementation should be avoided. I think running generators until they converge is doomed to fail even if its the naive implementation and probably much easier. I originally proposed it as one of the two solutions just so that it can be discussed, I honestly don't think its a good idea.

> Kinda feels either the design allows two generators to work off each others outputs with a fail safe to keep people from being ridiculous, or a strictly ordered system with a design that the Razor generator would almost immediately throw a wrench in.

The razor generator doesn't have to throw a wrench in it necessarily. The only point where the razor generator (I think) needs access to generated code is when it wants to use generated components. So the only non-supported use case I can think of off the top of my hat is generating new components on the fly within .razor files. And even then you _could_ order your generator to run before the razor generator just taking the .razor files directly as input.

> > and creating a new compilation. My gut feeling is that there are a lot of things you can do, but it won't be easy.
> 
> Any new compilation creates a linear additional cost on the entire pipeline. e.g. it's as if we're just recompiling the project 'one more time'.

Right. Worst case scenario is that we're going to have as many compilations as the dependency graphs longest path. This isn't something that can be helped directly. However my hope is that this is an initial penalty on first run and that subsequent runs would be able to shorten that path by making sure to only rerun generators whose inputs have changed, same as how incremental generators cache the results right now.

I think we all agree that we definitely will have to performance test any implementation to make sure that even a short chain of 2-3 generators will not triple compile times right off the bat and keep them tripled forever. Else that might indeed lead to a situation where working with chained generators becomes a chore.

**Comment by veler** [comment]
Hi there,

I'm hitting this issue too with a cross-platform project implying WASDK and some XAML.

Using [Uno Platform](https://github.com/unoplatform/uno), [CommunityToolkit.MVVM](https://github.com/CommunityToolkit/dotnet/) and [ReswPlus](https://github.com/DotNetPlus/ReswPlus) in the same project is pretty much impossible.

All these libraries I mentioned above have generators. Let's say I'm using something from `CommunityToolkit.MVVM` to generate the details of a property `Foo` in a ViewModel automatically and bind this property to the UI in the XAML. When doing so, I simply can't build my project because `Uno Platform` generators won't find `Foo` because it hasn't been generated by `CommunityToolkit.Mvvm` yet.

Workaround is to put all the XAML in a project A, all the ViewModels in a project B, all the RESW in a project C. But it unnecessarily complexify the solution architecture.

Now, by reading this thread, I do understand the big performance impact concern of sequentially running generators. I like the suggestion from @sanchez here 

> True. I hadn't considered that. You could group/nest sources thought right?
> 
> ```
> <Generators>
>   <Step>
>     <Source Type="GeneratorA" />
>     <Source Type="GeneratorB" />
>   </Step>
>   <Step>
>     <Source Type="GeneratorC" />
>   </Step>
> </Generators>
> ```
> 
> Each step being parallel within but being sequential adjacent. So A and B parallel together and C after
> 
> Edit: sorry realized I messed up your example a bit. Here A and B have no dependencies but C does for example

I wouldn't expect Roslyn to figure out itself which source generator depends on which other. I wouldn't expect source generator developers to set themselves what's their priority. I would instead expect to let the consumer of the source generator (me) defining in what order generator A B C should run. Documentation about this feature would have to clearly state that build performance may be drastically reduced depending on what we're doing. If it's clearly documented that slower build should be expected, I don't see why this wouldn't be an acceptable option.

**Comment by Elringus** [comment]
Another +1 here. I'm maintaining a [JS interop library](https://github.com/Elringus/DotNetJS) which provides various helpful tools absent in stock .NET (eg, generating bindings based on C# interfaces, type script definitions, etc) and it worked fine prior to .NET 7 where the JS interop layer has been changed and is now using source generators itself. My library also uses source generators, so it can't be migrated to the new interop. And the old interop is going to be deprecated in .NET 8. A dead end.

Source generators are spreading fast across the entire .NET runtime and there will be more such issues with community solutions in various areas (and eventually across .NET's own codebase, I believe); please consider giving the issue more attention.

**Comment by b-straub** [comment]
I like to mention the things you can do with SWIFT macros and property wrappers…
I feel the C# source generator approach is falling way behind when an elementary problem such as cascading generators is not solved.

**Comment by Jinjinov** [comment]
Can we expect this in .NET 8?

**Comment by jcouv** [comment]
> Can we expect this in .NET 8?

No, this is not planned (Backlog milestone).

**Comment by b-straub** [comment]
> > Can we expect this in .NET 8?
> 
> No, this is not planned (Backlog milestone).

I am quite amazed that when asked (Asp.Net team for AOT) even complex features like interceptors are implemented, but this topic, which is equally or even more important for AOT, remains in limbo for more than two years.

**Comment by CyrusNajmabadi** [comment]
@b-straub Interceptors are being investigated as they are likely to drive actual language changes, and we have to understand that space in order to move forward on it.  SGs are purely in the compiler, and the ability to run  before/after is well understood, but still lacks suitable solutions to the significant problems there.

**Comment by b-straub** [comment]
> but still lacks suitable solutions to the significant problems there.

The complexity has been understood, what about the proposed and considered solution?

https://github.com/dotnet/roslyn/issues/68779#issuecomment-1634960714

**Comment by CyrusNajmabadi** [comment]
> The complexity has been understood, what about the proposed and considered solution?

All generators are built off of an idea that they operate on a fully-representative versin of the world (called the 'Compilation').  In either SG v1 or SGv2, this means that in order to:

> choose another SG and pass my generated files for one more transform.

we would need to create a full new compilation, containing teh existing code and the code you generated, to then run the *other SG* and get its outputs.  Each of these transforms thus then adds a 'full Compilation' step which is enormously (and currently unacceptably) high cost. 

As nothing has changed with the costs, or the acceptability of those costs, this is still not a viable solution :(

**Comment by Elringus** [comment]
As an alternative, have you considered making all the generators used internally by .NET accessible as libraries? This way the users will be able to use them in their own generators, which (with a bit of additional hassle) will solve the issue as well.

The thing is, the more you utilize generators internally in .NET runtime, the less useful generators become for end users. You just keep breaking more and more extension points for the community. This is a really frustrating and worrying trend.

**Comment by CyrusNajmabadi** [comment]
You could certainly request that of the dotnet/runtime team :-)

**Comment by biohazard999** [comment]
I agree with @Elringus I don't think we need a Model where all SG are like in a tree depending on each other, which adds complexity and cost. I think a layered or phase approach is more than enough.

1. Roslyn built in
2. MS one's (like razor, json etc)
3. Community via nuget (they can't depend on others but the earlier ones
4. User code

Another idea would be to mark the generator in which phase they want to run, but of course this can lead to 'wild west' in the community/nuget space. But that's the same peoblem as with analyzers. And I think the community did a good job making sure the perf overhead is as minimal as possible.

**Comment by b-straub** [comment]
@Elringus

Would you take on the task of opening an issue with your idea in the runtime repository? Your suggestion would solve my case and I am sure many others as well.

**Comment by notour** [comment]
**are SG executed sequentially or in parrallel on the same code tree ?**

If SG are executed sequentially each SG could have an attribute expressing dependency. 
This way the order could be defined compile time like library dependencies.
This method allow more flexibility, order automatically through dependency tree, and error detection like cycling reference, ...

`[GeneratorSequenceDependency("....", Order.After | Order.Before)]`

If SG are in parallel it is more difficule and i understand the point of @CyrusNajmabadi

**Comment by Elringus** [comment]
> @Elringus
> 
> Would you take on the task of opening an issue with your idea in the runtime repository? Your suggestion would solve my case and I am sure many others as well.

I've actually already asked this via https://github.com/dotnet/runtime/discussions/87346#discussioncomment-6140606 for the new JS interop, and looks like they actually have plans on something like this, but they are long-term. I'm currently looking for alternative solutions, including switching out of C#/.NET entirely.

**Comment by b-straub** [comment]
> You could certainly request that of the dotnet/runtime team :-)

I don't think that language, compiler and runtime live on different planets considering that .NET development is largely driven by MS teams. Similar to how other teams can request new features from the language/compiler, it should be possible the other way around.

Sure, as a developer I can make a request to the runtime, but that won't be the same as a coordinated internal effort to address the problem that more and more AOT enhancements (JSON, RegEx, JSInteropt,...) are inaccessible to SG usage. This is not a small problem. Either the whole SG concept is not well enough thought out, or the negative effects that are now coming to light are not addressed actively enough.

**Comment by CyrusNajmabadi** [comment]
> I think as mentioned many times before the solution is to make dependencies explicit. That means that if one source generator has a dependency on another for processing their output it will run before that one. Obviously this means there can't be any loops and it needs to be a strictly acyclic graph.

This doesn't solve the primary concern that has been brought up since the beginning (and which i reiterated in my post).  Doing that approach fundamentally means producing N full compilations for *any* operation.  It's already problematic where N=2.  Extending that model further isn't currently tenable.

> Is there a specific reason why a compilation can't/isn't amended by the new source files instead of having to be done from scratch? 

There is no such concept as 'amended by the new source file'.  It's unclear what that would even mean or do.  If you add new stuff, presumably the intent is for it to be used, so it needs to be visible by the actual rest of the compilation being compiled.

Furthermore, roslyn is a big, immutable, graph.  You can't have a fully realized, immutable graph, then suddenly have more items in it.

Finally, a *core* principle of SGs is that if you took all the code generated, wrote it out to files, then recompiled those files with the original sources, that you would have the same output.  If 'amended by the new source files' doesn't impact the rest of hte code, this principle is broken. 

>  I don't expect there to be more than 2-3 passes needed in order to run the chain to its end.

Each pass is already too expensive.  Even the single pass we have today is teh source of major problems.  :-/

**Comment by Blackclaws** [comment]
> There is no such concept as 'amended by the new source file'. It's unclear what that would even mean or do. If you add new stuff, presumably the intent is for it to be used, so it needs to be visible by the actual rest of the compilation being compiled.

Yeah I've had a discussion on the C# Discord about this as well, I thought that this would somehow be feasible but there are quite a few places where it _would_ change the meaning of what's being done, such as partial classes, static constructors, replacing default constructors, static partial void methods, etc.

I was wondering whether there would be a way to cut down on the amount of work that has to be done for each pass and whether it wouldn't be somehow possible to restrict source generators that require other generators to a tighter set of inputs that could be generated much faster, though its a long shot. 

Even right now with SG2 there is a way to skip the run of source generators for cases where the inputs haven't changed. If that check could be made fast enough and if most of the time you would just skip running the intermediary source generator and use the cached generated source it should bring the total time closer to the time of a single run aside from the initial loading/first compile.

I think the problems with time you are referring to are mostly related to design time builds and not release builds?

**Comment by CyrusNajmabadi** [comment]
> Yeah I've had a discussion on the C# Discord about this as well, I thought that this would somehow be feasible but there are quite a few places where it would change the meaning of what's being done, such as partial classes, static constructors, replacing default constructors, static partial void methods, etc.

That's def one way that meaning changes.  But also, critically, the entire *model* of roslyn doesn't support this concept at all.  From the above post, we expose a "Full immutable graph" of everything.  So there's no way for that graph to be full and complete, but then have a section added to it. How would that work?  Either the graph is fully reachable and done, or you could observe that it was not complete and that it changed :)

We solve this in teh SG space by having each step be a *full, new, complete graph*.  Which, as you might imagine, is quite expensive.

Perhaps we could move the model away from thsi.  But now we're upending the very foundations of our compilation model in the first place.  Note that SGs have already upended this to some extent, and have come with huge costs to show for it.  Continuing down that path does not seem more viable, it seems less :(

**Comment by Blackclaws** [comment]
> Perhaps we could move the model away from thsi. But now we're upending the very foundations of our compilation model in the first place. Note that SGs have already upended this to some extent, and have come with huge costs to show for it. Continuing down that path does not seem more viable, it seems less :(

I think the question is what the future of SGs is supposed to be. As some have noted what is especially worrying for them is that parts of the runtime (such as JSInterop) are moving to a new source generator form while the "old-way" of doing things is getting deprecated. However any code coming from an SG does not have access to these features anymore. 

One way I could think of is to do what the Razor generator still allows, that is provide a fallback two-stage process, however that is basically what I'm proposing by saying: "opt-in to shooting yourself in the foot with a multistage process".

What they do is they limit their processing to a subset of files they know will impact their final result, this is of course much more difficult with source generators as a whole.  

Coming back around to the initial quote, if the runtime itself starts to make more and more use of SGs then it makes sense to me at least that there is _a_ way for other source generators to leverage that. It might be in the form a library that you can import (as discussed above), that you can then just pass your generated source to and that does its own postprocessing, or it could be in the form of some native roslyn chaining where SGs would only have to declare dependencies. 

I think the problem of not being able to run these generators off of each other is simply not going away and is only going to become larger as time goes on. Its a great language/compiler feature that SGs even exist so let's try to make them even more useful.

> Each pass is already too expensive. Even the single pass we have today is teh source of major problems. :-/

Personally, and this is of course just anecdotal evidence, I never had any issues even with larger projects that had multiple source generators, _as long_ as they were V2s or really fast V1s. Seeing the pass time doubled doesn't sound like a bad idea here, but again I don't have a concrete number to show around.

Are there concrete benchmarks you are referring to when talking about the single pass being a source of major problems?

**Comment by Jinjinov** [comment]
> Each of these transforms thus then adds a 'full Compilation' step which is enormously (and currently unacceptably) high cost.

> As nothing has changed with the costs, or the acceptability of those costs, this is still not a viable solution :(

> Each pass is already too expensive. Even the single pass we have today is teh source of major problems. :-/

@CyrusNajmabadi

Ok, I understand, it wouldn't be fast.

But it is not logical to talk about "enormously (and currently unacceptably) high cost" for a feature that would be opt-in for those who want it.

It's not like the solutions in the comments suggest that you force everyone to pay this high cost.

Is it not possible to provide the feature and let the users decide if they want to use it or not?

I would gladly run the whole build two times in a row, manually if need be, first `Build` and then `Build again with my SG` if that would mean that I could use the feature.

Why don't you let us decide if the high cost is acceptable for us or not? The rest of users could continue to use `Build`.

**Comment by CyrusNajmabadi** [comment]
> I would gladly run the whole build two times in a row, manually if need be, first Build and then Build again with my SG if that would mean that I could use the feature.

If you're manually doing things, you can accomplish this today.  Just write manual tools using the Roslyn api that do analysis, and just generate normal source files into some location that you then run with the normal compile.  

> Why don't you let us decide if the high cost is acceptable for us or not?

Because we have to design, implement, and maintain such a system for likely a decade or more.  And what happens in practice is customers say "hey, i started using X, and my performance tanked.  I blame you and you need to fix it.  or we stop paying" :)   And saying: "sorry, you opt'ed into using layered generators, so things will be bad" is not something we find people accept at all in these discussions.

**Comment by b-straub** [comment]
> Because we have to design, implement, and maintain such a system for likely a decade or more. And what happens in practice is customers say "hey, i started using X, and my performance tanked. I blame you and you need to fix it. or we stop paying" :) And saying: "sorry, you opt'ed into using layered generators, so things will be bad" is not something we find people accept at all in these discussions.

Strange argument, my sons MacBook Pro 14 compiled my large SG based project approx. 10 times faster than my two yours old high end Wintel Notebook (real measures). A decade is computer wise like a century.

**Comment by Jinjinov** [comment]
> If you're manually doing things, you can accomplish this today.

@CyrusNajmabadi

Can I really write a source generator that reads the C# classes generated from razor files in a Blazor project and adds a method if it is missing?

If I can do that manually today, I would really like to know how? :)

**Comment by montyclt** [comment]
> > If you're manually doing things, you can accomplish this today.
> 
> @CyrusNajmabadi
> 
> Can I really write a source generator that reads the C# classes generated from razor files in a Blazor project and adds a method if it is missing?
> 
> If I can do that manually today, I would really like to know how? :)

Like razor components are partial classes, I think you can generate an other part with this missing method.

**Comment by Blackclaws** [comment]
> > > If you're manually doing things, you can accomplish this today.
> > 
> > 
> > @CyrusNajmabadi
> > Can I really write a source generator that reads the C# classes generated from razor files in a Blazor project and adds a method if it is missing?
> > If I can do that manually today, I would really like to know how? :)
> 
> Like razor components are partial classes, I think you can generate an other part with the missing method.

You can only do that if you do the two step compile process for razor files. Normally razor files are compiled with an SG and you can't use the output in any other SG. With the two step process its possible. Its also possible to make a code behind .cs file that SGs can read as normal.

**Comment by biegehydra** [comment]
My problem with Blazor files right now is that I can't even analyze some parts of the razor file before the razor generator even runs.

For example, I really want to build a source generator that looks at all the parameters on a Blazor component. To mark all the Blazor components for source generation, the easiest thing to do would be.

```
@attribute [MyAttribute]
```

However, there is currently NO way that I am aware of to detect an attribute that is created this way.

The solution I've seen in some projects is to put the attribute on a partial class

```
[MyAttribute]
public partial class MyComponent
```

See the tests in this project https://github.com/excubo-ag/Generators.Blazor/blob/main/Tests_Blazor/SetParametersAsyncGenerator/EmptyClass.cs

However, for the use case of my source generator, having to create a code behind file just to apply the attribute simple isn't worth it.

**Comment by m-gasser** [comment]
@biegehydra
In one of our source generators we consume all Razor files as AdditionalTexts (with the same filter criteria that [RazorSourceGenerator](https://github.com/dotnet/razor-compiler/blob/65ca79041826071d4f31f84e7cb442b2642cfbd3/src/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs#L36) is using). Then we parse those files with a regex instead of consuming syntax or semantic nodes. This way we can handle the most common cases. Because regex is obviously not bullet proof, since it just parses the raw text, we also have an analyzer, which we execute for all "_razor.g.cs" and "_cshtml.g.cs" files. The analyzer checks if what we generated based on the regex matches what we expect. Otherwise it errors out with a link to the relevant known issues page.
Maybe you are lucky as we were and you can get all the information you need from a regex.

This is actually not the only workaround we had to do. In another case we cannot get the type of an expression, if it is dependent on a type generated by another source generator. We then get an IInvalidOperation and just take the assumption that we have a conflict with another source generator, instead of doing the proper checks.

Whilst the workarounds we implemented are quite messy, using many source generators works now quite nicely for us. I am however not sure if it stays this way, since Microsoft is releasing more and more source generators. It is possible that we get in a situation, where we cannot use some of those source generators.

So while "Source generators have to be independent" is a quite reasonable limitation, in practice it unfortunately might not show up as limitation but as buggy behavior instead. In our case it even shows up as silent fails, which is especially nasty.

That this limitation can show up as buggy behavior might actually be a chance. It means, that a solution that is not perfect would still be a big improvement. So instead of creating more full compilations than the current two, generators could get an "estimated" incremental semantic model of just the code added by other source generators. The symbols in this incremental compilation would bind primarily to other symbols in the incremental compilation and secondarily to the initial full compilation. All other code (that was not generated by a source generator) will not be recompiled or rebound. The result is likely quite wrong according to the rules of the language, but the hypothesis is, that it will be good enough for source generators to reason about. Since such an increment considers only the code generated by the source generators, it should be reasonably small and thus fast to create. Source generators would be able to refine the code they already generated and this code could be fed back into all generators, until this converges (which is usually quite soon, as others have already noted). After this step, all the generated code can be feed into the second full compilation, where everything gets correctly bound. If a source generator did generate something wrong based on the incorrect immediate semantic model, this can lead to a compile error or a silent fail. Either way, we are not worse off as we are now. Currently we get an not very helpful error when trying to use the Regex source generator in a Razor page or we a silent fail, as is the case with one of our source generators. So the way it can fail would stay the same, it would just be much less likely to happen.

**Comment by Cybrosys** [comment]
Hello,

_To avoid creating one more proposal/issue I decided to comment here and you all can tell me if I'm wrong or should create a new proposal/issue._

I am currently using the .NET Community Toolkit for MVVM and `[ObservableProperty]` to generate public properties for fields marked with this attribute. I also have a wish to use Mapperly, a source generator for generating object mappings. 

However as I noticed, Mapperly is not able to detect the public properties that are generated by the MVVM Community Toolkit. Since I assumed this was an issue with Mapperly I opened an issue which was closed as this was an issue with Roslyn:
https://github.com/riok/mapperly/issues/883

So, would the work related to this feature, enable Mapperly to find the public properties generated by the Toolkit?

**Comment by mnelsonwhite** [comment]
ordering generators is necessary to allow it to be useable in non-trivial cases.

It doesn't make sense to me to specify ordering by use of an attribute in the SG code. This is a problem that should be solved by configuration, just as we do when trying to depend on packages of multiple differing versions.

Just as before, we allow users to configure how the project should work with their knowledge of how the code with utilise them.

Therefore the simplest and best solution at this stage is to introduce some way of specifying the order in the project file.

The suggested syntax shown here by @sanchez seems like a sensible option, so long as it doesn't require knowledge of all SGs which are run, and that those which are not specified by a step, still run in any order.

> > ```
> > <Generators>
> >   <Step>
> >     <Source Type="GeneratorA" />
> >     <Source Type="GeneratorB" />
> >   </Step>
> >   <Step>
> >     <Source Type="GeneratorC" />
> >   </Step>
> > </Generators>
> > ```

**Comment by ThaDaVos** [comment]
Just to give my opinion, I think the ordering should be determined inside the Source Generator itself, as the developer of the generator knows best what the order should be.

My idea would be to include the ordering in the `Initialize` method of the source generator with two methods on the `GeneratorPostInitializationContext`, as this already supports adding sources to the compilation - one method to mark the current Generator to always run before another and one to mark the current Generator to always run after another - this way, the ordering is determined in the current generator and doesn't need involvement/changes in other generators, maybe as follows:

```C#
public void RunBefore(string generatorAssemblyName);
public void RunAfter(string generatorAssemblyName);
```

**Comment by Youssef1313** [comment]
@ThaDaVos The main problem here is not really the user-facing design of the feature (i.e, how users will use the feature). The problem here is mostly how to implement the feature in a performant manner. The performance of source generators is very important.

But, aside from that, I think there are scenarios where *consumers* of source generators want to define the ordering. In that case, my opinion would be that MSBuild properties/items are more suitable. Having the ordering defined at build-time with MSBuild allows the generator authors to define ordering, and also allowing consumers to adjust that ordering.

**Comment by weltkante** [comment]
> why would a consumer of the generators want to change their order

Because its impossible for a developer of a generator to know _all_ other generators, including future ones, that they may depend on. If a source generator acts on e.g. UI bindable properties and generates the bindings (think x:Bind) I may want to run it after any source generator providing UI bindable properties, and that is an open list, it needs to run after an unknown (at the time of writing the generator) number of other generators.

Anyways, the whole discussion is kinda moot, the performance is the main problem, as the previous discussion clearly concluded. Roslyn would have to switch its compilation data structures to an incremental data structure to even think about doing chained generators. [[see discussion above]](https://github.com/dotnet/roslyn/issues/57239#issuecomment-1662470655)

**Comment by Blackclaws** [comment]
I'm wondering whether making a new category of Source Generator: "Source Postprocessor" would alleviate a few/most issues people have here.

A lot of generators fall broadly into two categories:

- Those adding classes/behaviour to existing classes (System.Text.Json, INotifyPropertyChanged, etc.)
- Those using existing classes/behaviour to drive their own (Mediator, etc.)

The second class is usually the one that you'd want to consume the output of the first class while any dependency on another source generator they themselves have they could resolve by piping their own generated source through the other generator manually (which I think is something you can actually do though YMMV).

The first class would stay as normal Source Generators while the second class would be Source Postprocessors, strictly running as their own second/third step after all normal Source Generators are already done with their work. We could also make it a rule that they can't participate in Design Time builds or something like that if we're worried about the IDE experience.

**Comment by sharwell** [comment]
Not sure if I mentioned this above, but one area where it feels like the system is lacking currently is the ability to generate source from all the _inputs_ to a `Compilation` prior to the creation of the `Compilation` itself.

* Metadata references (potentially including the `IAssemblySymbol` for those references)
* Additional files
* Analyzer config information

This step would not increase binding work within the compiler, which is a primary concern for layering source generators.

**Comment by m-gasser** [comment]
> generate source from all the _inputs_ to a `Compilation` prior to the creation of the `Compilation` itself.
I like that idea! We could use this for 2 of our source generators. If the Razor source generator would use that mechanism too, this would solve most of our problems.

**Comment by SongOfYouth** [comment]
Maybe likes the `before` or `after` property of msbuild task?

**Comment by chsienki** [comment]
> I like that idea! We could use this for 2 of our source generators. If the Razor source generator would use that mechanism too, this would solve most of our problems.

Razor wouldn't be able to use this mechanism, as it has to combine both non-source files (Razor files) but also items in the compilation. For example you can declare a tag helper in source that is used in the razor file, so razor is more than strictly a pre-processor. 

I'm not against an improved post-init step that includes access to things like metadata refs though as a general proposal.

**Comment by montyclt** [comment]
Stay aware as Blazor problems, https://github.com/dotnet/aspnetcore/issues/32172 too.

**Comment by codymullins** [comment]
Is there any kind of official update for this issue? It would be nice to know what kind of time horizon to expect for this, if there is one. We have some really cool things we want to do with SGs, but this particular issue with the razor source generator makes the dev experience less desirable.

**Comment by CyrusNajmabadi** [comment]
@codymullins There is no update on this.  The design is that generators see the initial compilation, and produce results based on that.  The final results of all generators are then combined, along with the user's original source to produce the final compilation.

**Comment by CyrusNajmabadi** [comment]
>  It should have been expected that more than one library would use this feature and one would end up depending on the other.

The desire is understood.  But it opens up too many problematic issues.  Generators *already* are very difficult to get right (from a performance/incrementality perspective) even with the current design.  Allowing even more ways for plugins to significantly degrade the compiler invariants doesn't help that at all.

>  It's likely that it was a design flaw. 

Sometimes it's not possible to get high flexibility *and* high performance.  Limits need to be made to ensure the health/viability of the feature as a whole.

**Comment by weltkante** [comment]
Is there a tracking issue for making the compilation incremental? It has been a while since this appeared in the discussion and I didn't re-read, but I think the compilation being replaced every build was a major source of the performance issues?

It's a solved problem making mutable object graphs immutable allowing incremental updates from the latest version, similar to how VS TextBuffers allow edits only on the newest version. I think that would be sufficient for Roslyn as it likely doesn't need to make new compilations based on anything but the previous compilation.

(For people who don't know how that works, its usually called "persistent data structures" not "immutable" and works by reserving a variable in each object to record how a future version changed, instead of immediately make a copy when a new version is required. I think the main paper explaining this technique was called "making data structures persistent" but I may be mistaken, its been a while I've looked into this topic.)

I understand if you have different priorities at the moment, but rejecting that this is a solvable problem seems ... wrong?

**Comment by Cybrosys** [comment]
I'm currently using one SG to generate observable properties based on fields (.NET MAUI and MVVM bindings). I would like to also use Mapperly, a SG for mapping from one type to another. However, it can't generate mappings between the DTOs and models because the models' properties are generated.

That specific scenario could be worked around if ever support for partial properties are implemented [6420](https://github.com/dotnet/csharplang/issues/6420) and the MVVM SG could work against those instead of fields (or both).

But as things are now, I can only use the SG for observable properties and write any mapping logic manually.

**Comment by mikernet** [comment]
@Cybrosys FWIW, Uno source generators work around this by "detecting" community toolkit generated properties from the backing fields, and then use those props. You could do something similar with mapperly, i.e. fork it into MvvmMapperly or whatever, and add support for detecting that a prop will be generated from the fields.

**Comment by Cybrosys** [comment]
> @Cybrosys FWIW, Uno source generators work around this by "detecting" community toolkit generated properties from the backing fields (and their annotations), and then use those props. You could do something similar with mapperly, i.e. fork it into MvvmMapperly or whatever, and add support for detecting that a prop will be generated from the fields.

Thank you for the suggestion. I was thinking along those same lines as I've had to write a couple of other SG for MAUI, but in the end I went with the pragmatic approach of making my models be the DTOs instead. Sometimes you just need to get things done 😉

**Comment by HavenDV** [comment]
@Cybrosys For some cases, next approach may work - move part of the project to a library in which one of the generators will work. And apply the second generator to the project using the library.
Of course, you can also run multi-level compilation by adding a little MSBuild code, but this cannot be applied as a general solution because it will create huge performance problems

**Comment by Cybrosys** [comment]
> @Cybrosys For some cases, next approach may work - move part of the project to a library in which one of the generators will work. And apply the second generator to the project using the library. Of course, you can also run multi-level compilation by adding a little MSBuild code, but this cannot be applied as a general solution because it will create huge performance problems

Now this was a very interesting workaround, using multiple projects.

**Comment by CyrusNajmabadi** [comment]
> Is there a tracking issue for making the compilation incremental?

Compilations are already incremental in all the ways that are safe to do so based on the design of the language.  Tons of changes though simply cannot be incremental, and have deep implications on what must be reanalyzed.

SGs also have an viable incremental path that we recommend.  HOwever, even with this path, the cost of doing incremental compilation is still very high, and continually a problem for performance for higher layers of our stack.

> It has been a while since this appeared in the discussion and I didn't re-read, but I think the compilation being replaced every build was a major source of the performance issues?

The compilation is replaced, but it happens in an incremental fashion.  That said, incremental doesn't mean we still meet the perf needs of all the layers that sit on top of the compielr.

> It's a solved problem making object graphs immutable while still allowing incremental updates from the latest version, similar to how VS TextBuffers allow edits only on the newest version. I think that would be sufficient for Roslyn as it likely doesn't need to make new compilations based on anything but the previous compilation.

We already do this.

> (For people who don't know how that works, its usually called "persistent data structures" not "immutable" and works by reserving a variable in each object to record how a future version changed, instead of immediately making a copy when a new version is required. 

We already do this.

> I think the main paper explaining this technique was called "making data structures persistent" but I may be mistaken, its been a while I've looked into this topic.)

You may be thinking of "Purely Functional Data Structures" by Chris Okasaki.  I can assure you, we're already using those designs and techniques.  But we still have extreme latency goals as well, and even though have trouble.

> > Sometimes it's not possible to get high flexibility and high performance.

> I understand if you have different priorities at the moment, but rejecting that this is a solvable problem seems ... wrong?

No one has rejected it as a solvable problem.  WHat's been said is that *until* we have a design that solves the problem, it's much more difficult to get something in here due to the devastating impact it could have.

**Comment by PeterHevesi** [comment]
I have found this awesome tool called [Metalama](https://www.postsharp.net/metalama) based on Roslyn (they have their own fork of Roslyn), that can do AOP, but beside that, can also generate code (new fields and properties, new methods and even classes) and you can specify order of those "Metalama aspects" there! While not needing any additional compilations !
It can't solve the problem with built-in .NET's Roslyn generators, as all "Metalama aspects" run after Roslyn source generators..
But at least while using Metalama, you can specify order between those "aspects" without performance killing !

**Comment by weltkante** [comment]
> I think only the immutable syntax/operation tree generated by a previous generator is sufficient to write a source generator based on another generator's output

Some (many?) generators eventually access the semantic model, which isn't immutable yet, and is the expensive part of chaining generators. It was mentioned previously further up this discussion.

**Comment by b-straub** [comment]
For me this discussion goes somewhat off-topic. While it might be desirable chaining arbitrary SG together, I would be more than happy, if my SG could use the increasing number of runtime SGs such as the one for JSON. Especially for AOT aware development this is a major issue.

It's one thing implementing something experimental, that might turn out being not the right path to the future, but something totally different increasingly using this feature across the entire runtime and still complaining about the limitations of the approach.

**Comment by weltkante** [comment]
> only the syntax should be sufficient as the 'common model' between generators. If you're building on top of another, you're already coupled to its exact output, so there's no reason to do semantic analysis on the generated code which almost certainly will have a predefined structure.

You often need names bound to something to know their meaning; most commonly used for types, for example to check if the type returned by a property is a class or struct, or to check for attributes on the type. There is a lot that you can't do with just syntax. There's a reason why even the builtin code generators require the semantic model.

> For example, imagine you want to add some partial definitions on types generated by another generator. The only thing you would need is the syntax, the semantics are completely predictable at that point.

You probably need to know the return type. Sure, its predictable, but then you replicate the binding of names to symbols that the semantic model does.

**Comment by codymullins** [comment]
There's a lot of discussion going on, which is great, but I'd love a clear understanding of what specifically needs to happen for this item to be under active investigation. As it stands, it feels this issue is not a priority at all for the .NET team, while the community continually circles back to wanting this feature to be expanded. @CyrusNajmabadi or another maintainer, if you could help us summarize where we are here - that would be wonderful.

**Comment by CyrusNajmabadi** [comment]
Nothing has changed here. The rules and restrictions around generators have been clearly delineated and we have no plans on changing them currently.

**Comment by codymullins** [comment]
So should this be marked as "Closed, won't do"?

**Comment by CyrusNajmabadi** [comment]
> Thanks. So should this be marked as "Closed, won't do"?

Not for me.  Not planning to do anything now doesn't mean we won't do something in the future.

**Comment by kzu** [comment]
My advise to anyone coming here: the team does not seem interested in solving this.
The only workaround possible is to re-run the generators that should be aware of your code, from your own generator.

For example: 
1. Orleans (thankfully!) created a separate package with its [CodeGenerator](https://www.nuget.org/packages/Microsoft.Orleans.CodeGenerator) which you can invoke from your own source generator. Theirs mostly just calls it out too [so you can copy/paste a bit](https://github.com/dotnet/orleans/blob/main/src/Orleans.CodeGenerator/OrleansSourceGenerator.cs#L73-L74) and be done.
2. Razor provides the [language](https://www.nuget.org/packages/Microsoft.AspNetCore.Razor.Language) package which you can also invoke manually after you generated your stuff, like:

```csharp
        var generated = new HashSet<string>(texts.Where(x => x.Path.EndsWith(".g.cs")).Select(x => x.Path));
        var razors = new HashSet<string>(texts.Where(x => x.Path.EndsWith(".razor")).Select(x => x.Path));
        var options = compilation.SyntaxTrees.FirstOrDefault()?.Options as CSharpParseOptions;
        
        var fs = RazorProjectFileSystem.Create(project.ProjectDirectory);
        var engine = RazorProjectEngine.Create(
            RazorConfiguration.Default, fs,
            builder => builder.SetRootNamespace(project.RootNamespace));

        var trees = fs.EnumerateItems(project.ProjectDirectory)
            // Filter out files we find that aren't razor files currently passed in for generation for whatever reason, via targets
            .Where(x => razors.Contains(Path.GetFullPath(x.CombinedPath)))
            .Select(x => CSharpSyntaxTree.ParseText(engine.Process(x).GetCSharpDocument().GeneratedCode, 
                path: Path.GetFullPath(x.CombinedPath), options: options))
            .ToImmutableArray();

        var razored = compilation.AddSyntaxTrees(trees);
```

(this allows you to inspect the types that razor would generate, since you cannot run *after* it and see its output).

Pretty sure for the razor generator you could end up doing something similar, or eventually just invoke their generator manually by loading the assembly from the analyzer assemblies you can pass to your generator as additional files and do a bit of reflection, etc.

And so on. Eventually, all of this will become so slow when everyone is doing it, that they might come back to this and finally see the importance of the use case. And then they will have to go chasing everyone who was forced to work around this incomplete design and beg for them to stop doing it and use the "improved design" then. Pretty sure there will be analyzers trying to stop you from just getting the job done, before they just give up and fix it. :)

It's not really hard to imagine the myriad scenarios where this is necessary. I alone was faced by the above two myself in about a year. It's inevitable.

**Comment by CyrusNajmabadi** [comment]
> My advise to anyone coming here: the team does not seem interested in solving this.

The team is interested.  But no one (internal or external) has been able to find a viable solution.  We can't move forward without an actual design that will work.

**Comment by CyrusNajmabadi** [comment]
> before they just give up and fix it. :)

How do you suggest fixing it?  We have no magic bullet at our disposal.

**Comment by kzu** [comment]
I'm confident the day when one source-generated feature **in the BCL** depends on another source-generated feature, a viable solution will be discovered. I'm just hoping that happens sooner than later, but I know it will happen.

**Comment by CyrusNajmabadi** [comment]
> I'm just hoping that happens sooner than later, 

If a suitable design can be presented, we'd be very interested.  That is true whether that design comes from BCL partners, or the community.  Neither has been able to come up with a way to address the problems presented so far.

**Comment by kzu** [comment]
I'm pretty sure when the heat comes from BCL partners, "a suitable design will present itself" 😉

**Comment by CyrusNajmabadi** [comment]
> I'm pretty sure when the heat comes from BCL partners, "a suitable design will present itself" 😉

That hasn't happened so far.  The opposite happened.  We created a limited, but workable design, and we pushed the entirety of the bcl onto it.  That's what v2 generators were. Exactly because it's not tenable to try to build on to if a bad foundation.

**Comment by CyrusNajmabadi** [comment]
If a suitable design is presented, we will consider it. But so far that hasn't happened. We have not come up with one. Our partners have not come up with one. The community has not come up with one.  People need to stop thinking there's some magic bullet here, or that we're not doing it because we don't see the value.  We're not doing it because no one knows how.  If that changes we can take next steps.
